### PR TITLE
AMDGPU: Make bf16/v2bf16 legal types

### PR DESCRIPTION
**Pull request by arsenm**
There are some intrinsics are using i16 vectors in place of bfloat vectors.
Move towards making bf16 vectors legal so these can migrate. Leave the
larger vectors for a later change.

Depends #76213 #76214

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -3901,6 +3958,26 @@ SDValue SITargetLowering::lowerPREFETCH(SDValue Op, SelectionDAG &DAG) const {
   return Op;
 }
 
+// Work around DAG legality rules only based on the result type.
+SDValue SITargetLowering::lowerFP_EXTEND(SDValue Op, SelectionDAG &DAG) const {
+  bool IsStrict = Op.getOpcode() == ISD::STRICT_FP_EXTEND;
```
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:3963]: tiny tiny nit: put parentheses around the initializer - `=` and `==` on each side of an expression is just a bit confusing at first glance

File: llvm/test/CodeGen/AMDGPU/fmed3-cast-combine.ll [hunk 2]
```diff
@@ -1021,13 +1021,19 @@ define half @fmed3_f32_fpext_bf16(bfloat %arg0, bfloat %arg1, bfloat %arg2) #1 {
 ; GFX8-LABEL: fmed3_f32_fpext_bf16:
 ; GFX8:       ; %bb.0:
 ; GFX8-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX8-NEXT:    v_lshlrev_b32_e32 v0, 16, v0
```
Comment (Pierre-vh) [inline llvm/test/CodeGen/AMDGPU/fmed3-cast-combine.ll:1024]: Why do we have these now?

**Review by Pierre-vh — COMMENTED** [review]
Test changes are a bit hard to review so I didn't look at all of them

Do we need a GISel equivalent of this patch?

**Comment by arsenm** [comment]
No, legal types are more of a DAG concept

**Review by rampitec — APPROVED** [review]
LGTM, although I do not believe we can finally make bfloat16 legal in an FE now, we will still have a lot of emulation for many operations. I.e. if the final plan is to switch from i16 to bf16 in intrinsics and builtins, that is hardly possible. At best we will need to start with overloaded interfaces supporting both.

**Comment by arsenm** [comment]
That's what the comment says but everything I tried works by promotion to float. I expanded the test to cover just about everything

**Pull request merged**